improver/nbhood/nbhood.py

```diff
@@ -383,19 +391,34 @@ def _do_nbhood_sum(
                     _xstart,
                     _xstop,
                 )
+        square_buffer_kwargs = {"mode": "constant"}
         if size != data.size:
             # Determine default array for the extremes around the edges, or everywhere
             if isinstance(when_all_extremes, np.ndarray):
                 untrimmed = when_all_extremes.astype(data.dtype)
             else:
                 untrimmed = np.full(data_shape, when_all_extremes, dtype=data.dtype)
-        if size:
-            # Trim to the calculated box
-            data = data[ystart:ystop, xstart:xstop]
+            if (
+                self.neighbourhood_method == "square"
+                and min(ystart, xstart, data_shape[-2] - ystop, data_shape[-1] - xstop)
+                >= half_nb_size + 1
+            ):
+                # Trim to the calculated box plus padding
+                # and rely on boxsum to return the right sized array
+                data = data[
+                    ystart - half_nb_size - 1 : ystop + half_nb_size,
+                    xstart - half_nb_size - 1 : xstop + half_nb_size,
+                ]
+                square_buffer_kwargs = {}
+            else:
+                # Trim to the calculated box without padding and let boxsum do its padding.
+                # Circular neighbourhoods will ignore the square_buffer_kwargs.
+                data = data[ystart:ystop, xstart:xstop]
 
-            # Calculate neighbourhood totals for input data.
+        if size:
+            # Calculate neighbourhood totals for input data if it isn't all constant.
             if self.neighbourhood_method == "square":
-                data = boxsum(data, self.nb_size, mode="constant")
+                data = boxsum(data, self.nb_size, **square_buffer_kwargs)
```

Comment:
The only issue with your approach is that I didn't see your suggestion and I didn't think about this myself until I was tucking in to my lunch. I was also tempted by the satisfaction of trimming the original data to exactly the right size, rather than cutting big bits off and then sticking small bits back on again.

I've put your suggestion in, taken most of my changes out, kept the new tests, and made one little change to skip the optimisation for complex numbers.